archunit/src/main/java/com/tngtech/archunit/library/modules/ArchModules.java

```diff
@@ -0,0 +1,556 @@
+        }
+
+        /**
+         * Allows to derive the {@link ArchModule.Descriptor} from the {@link JavaClass root class} that induced the respective {@link ArchModule}.
+         *
+         * @param descriptorCreator A function describing how to derive the {@link ArchModule.Descriptor} from the respective
+         *                          {@link ArchModule.Identifier} and {@link JavaClass root class}
+         * @param <D>               The specific subtype of {@link ArchModule.Descriptor}
+         * @return A fluent API to further customize how to create {@link ArchModules}
+         */
+        @PublicAPI(usage = ACCESS)
+        public <D extends ArchModule.Descriptor> WithGenericDescriptor<D> describeModuleByRootClass(RootClassDescriptorCreator<D> descriptorCreator) {
+            return describeBy((identifier, __) -> descriptorCreator.create(identifier, identifierAssociation.getRootClassOf(identifier)));
+        }
+
+        static CreatorByRootClass from(Predicate<? super JavaClass> rootClassPredicate) {
+            return new CreatorByRootClass(new RootClassIdentifierAssociation(rootClassPredicate));
+        }
+
+        private static class RootClassIdentifierAssociation implements IdentifierAssociation {
+            private final Map<String, Identifier> packageToIdentifier = new HashMap<>();
+            private final Map<Identifier, JavaClass> identifierToRootClass = new HashMap<>();
+            private final Predicate<? super JavaClass> rootClassPredicate;
+
+            private RootClassIdentifierAssociation(Predicate<? super JavaClass> rootClassPredicate) {
+                this.rootClassPredicate = rootClassPredicate;
+            }
+
+            @Override
+            public void init(Collection<JavaClass> allClasses) {
+                allClasses.stream().filter(rootClassPredicate).forEach(rootClass -> {
+                    packageToIdentifier.keySet().forEach(pkg -> {
+                        if (packagesOverlap(pkg, rootClass.getPackageName())) {
+                            throw new IllegalArgumentException(String.format(
+                                    "modules would overlap in '%s' and '%s'", pkg, rootClass.getPackageName()));
+                        }
+                    });
+                    Identifier identifier = Identifier.from(rootClass.getPackageName());
+                    packageToIdentifier.put(rootClass.getPackageName(), identifier);
+                    identifierToRootClass.put(identifier, rootClass);
+                });
+            }
+
+            private boolean packagesOverlap(String firstPackageName, String secondPackageName) {
+                return packageContains(firstPackageName, secondPackageName) || packageContains(secondPackageName, firstPackageName);
+            }
+
+            private boolean packageContains(String parentPackage, String childPackage) {
+                return childPackage.equals(parentPackage) || childPackage.startsWith(parentPackage + ".");
```

Comment:
I know what you mean, but since for one side we really just have the package name I wouldn't know how it should solve this case completely? I mean, we could check `rootClass.getPackage().containsPackageWithName(pkg)`, but how would the other way round look like if `pkg` is just a string? And since `pkg` is derived from the module declaration (and thus is no `JavaPackage` I don't know how a method on `JavaPackage` could solve this then?

Reply:
As discussed, we're gonna keep this for now. If we should improve `JavaPackage` later on to cover this case we can still revisit this.